















<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2020-12-18 15:56:59 +0000">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
     <link rel="stylesheet" type="text/css" href="../assets/css/fonts.css" />
    <link rel="license" href="#license-info" />

    



    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/dc/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/dc/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/dc/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/dc/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/dc/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/dc/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/dc/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/dc/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/dc/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Data Carpentry - Intro to R and RStudio for Genomics"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/dc/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/dc/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/dc/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/dc/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/dc/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->
  <title>
  R Basics continued - factors and data frames &ndash; Intro to R and RStudio for Genomics
  </title>

  </head>
  <body>
    


<div class="panel panel-default life-cycle">
  <div id="life-cycle" class="panel-body alpha">
    This lesson is in the early stages of development (Alpha version)
  </div>
</div>





    <div class="container">
      
















  
  










<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="http://datacarpentry.org" class="pull-left">
        <img class="navbar-logo" src="../assets/img/dc-icon-black.svg" alt="Data Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../index.html">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="../CODE_OF_CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="../setup.html">Setup</a></li>

        
        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-introduction/index.html">Introducing R and RStudio IDE</a></li>
            
            
            <li><a href="../02-r-basics/index.html">R Basics</a></li>
            
            
            <li><a href="../03-basics-factors-dataframes/index.html">R Basics continued - factors and data frames</a></li>
            
            
            <li><a href="../04-dplyr/index.html">Aggregating and Analyzing Data with dplyr</a></li>
            
            
            <li><a href="../05-data-visualization/index.html">Data Visualization with ggplot2</a></li>
            
            
            <li><a href="../XX-knitr-markdown/index.html">Producing Reports With knitr</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="../aio.html">All in one page (Beta)</a></li>
          </ul>
        </li>
        
	

	
	
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="../reference.html">Reference</a></li>
            
            
            <li><a href="../about/index.html">About</a></li>
            
            
            <li><a href="../discuss/index.html">Discussion</a></li>
            
            
            <li><a href="../figures/index.html">Figures</a></li>
            
            
            <li><a href="../guide/index.html">Instructor Notes</a></li>
            
          </ul>
        </li>
	

	
        <li><a href="../LICENSE.html">License</a></li>
	
	
	<li><a href="/edit//_episodes_rmd/03-basics-factors-dataframes.Rmd" data-checker-ignore>Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
	
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>

      






      


























  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../02-r-basics/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Intro to R and RStudio for Genomics</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../04-dplyr/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">R Basics continued - factors and data frames</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 60 min
      <br/>
      <strong>Exercises:</strong> 30 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How do I get started with tabular data (e.g. spreadsheets) in R?</p>
</li>
	
	<li><p>What are some best practices for reading data into R?</p>
</li>
	
	<li><p>How do I save tabular data generated in R?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Explain the basic principle of tidy datasets</p>
</li>
	
	<li><p>Be able to load a tabular dataset using base R functions</p>
</li>
	
	<li><p>Be able to determine the structure of a data frame including its dimensions and the datatypes of variables</p>
</li>
	
	<li><p>Be able to subset/retrieve values from a data frame</p>
</li>
	
	<li><p>Understand how R may coerce data into different modes</p>
</li>
	
	<li><p>Be able to change the mode of an object</p>
</li>
	
	<li><p>Understand that R uses factors to store and manipulate categorical data</p>
</li>
	
	<li><p>Be able to manipulate a factor, including subsetting and reordering</p>
</li>
	
	<li><p>Be able to apply an arithmetic function to a data frame</p>
</li>
	
	<li><p>Be able to coerce the class of an object (including variables in a data frame)</p>
</li>
	
	<li><p>Be able to import data from Excel</p>
</li>
	
	<li><p>Be able to save a data frame as a delimited file</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h2 id="working-with-spreadsheets-tabular-data">Working with spreadsheets (tabular data)</h2>

<p>A substantial amount of the data we work with in genomics will be tabular data,
this is data arranged in rows and columns - also known as spreadsheets. We could
write a whole lesson on how to work with spreadsheets effectively (<a href="https://datacarpentry.org/organization-genomics/">actually we did</a>). For our
purposes, we want to remind you of a few principles before we work with our
first set of example data:</p>

<p><strong>1) Keep raw data separate from analyzed data</strong></p>

<p>This is principle number one because if you can’t tell which files are the
original raw data, you risk making some serious mistakes (e.g. drawing conclusion
from data which have been manipulated in some unknown way).</p>

<p><strong>2) Keep spreadsheet data Tidy</strong></p>

<p>The simplest principle of <strong>Tidy data</strong> is that we have one row in our
spreadsheet for each observation or sample, and one column for every variable
that we measure or report on. As simple as this sounds, it’s very easily
violated. Most data scientists agree that significant amounts of their time is
spent tidying data for analysis. Read more about data organization in
<a href="https://datacarpentry.org/organization-genomics/">our lesson</a> and
in <a href="https://www.jstatsoft.org/article/view/v059i10">this paper</a>.</p>

<p><strong>3) Trust but verify</strong></p>

<p>Finally, while you don’t need to be paranoid about data, you should have a plan
for how you will prepare it for analysis. <strong>This a focus of this lesson.</strong>
You probably already have a lot of intuition, expectations, assumptions about
your data - the range of values you expect, how many values should have
been recorded, etc. Of course, as the data get larger our human ability to
keep track will start to fail (and yes, it can fail for small data sets too).
R will help you to examine your data so that you can have greater confidence
in your analysis, and its reproducibility.</p>

<blockquote class="callout">
  <h2 id="tip-keeping-you-raw-data-separate">Tip: Keeping you raw data separate</h2>

  <p>When you work with data in R, you are not changing the original file you
loaded that data from. This is different than (for example) working with
a spreadsheet program where changing the value of the cell leaves you one
“save”-click away from overwriting the original file. You have to purposely
use a writing function (e.g. <code class="language-plaintext highlighter-rouge">write.csv()</code>) to save data loaded into R. In
that case, be sure to save the manipulated data into a new file. More on this
later in the lesson.</p>
</blockquote>

<h2 id="importing-tabular-data-into-r">Importing tabular data into R</h2>
<p>There are several ways to import data into R. For our purpose here, we will
focus on using the tools every R installation comes with (so called “base” R) to
import a comma-delimited file containing the results of our variant calling workflow.
We will need to load the sheet using a function called <code class="language-plaintext highlighter-rouge">read.csv()</code>.</p>

<blockquote class="challenge">
  <h2 id="exercise-review-the-arguments-of-the-readcsv-function">Exercise: Review the arguments of the <code class="language-plaintext highlighter-rouge">read.csv()</code> function</h2>

  <p><strong>Before using the <code class="language-plaintext highlighter-rouge">read.csv()</code> function, use R’s help feature to answer the
following questions</strong>.</p>

  <p><em>Hint</em>: Entering ‘?’ before the function name and then running that line will
bring up the help documentation. Also, when reading this particular help
be careful to pay attention to the ‘read.csv’ expression under the ‘Usage’
heading. Other answers will be in the ‘Arguments’ heading.</p>

  <p>A) What is the default parameter for ‘header’ in the <code class="language-plaintext highlighter-rouge">read.csv()</code> function?</p>

  <p>B) What argument would you have to change to read a file that was delimited
by semicolons (;) rather than commas?</p>

  <p>C) What argument would you have to change to read file in which numbers
used commas for decimal separation (i.e. 1,00)?</p>

  <p>D) What argument would you have to change to read in only the first 10,000 rows
of a very large file?</p>

  <blockquote class="solution">
    <h2 id="solution">Solution</h2>

    <p>A) The <code class="language-plaintext highlighter-rouge">read.csv()</code> function has the argument ‘header’ set to TRUE by default,
this means the function always assumes the first row is header information,
(i.e. column names)</p>

    <p>B) The <code class="language-plaintext highlighter-rouge">read.csv()</code> function has the argument ‘sep’ set to “,”. This means
the function assumes commas are used as delimiters, as you would expect.
Changing this parameter (e.g. <code class="language-plaintext highlighter-rouge">sep=";"</code>) would now interpret semicolons as
delimiters.</p>

    <p>C) Although it is not listed in the <code class="language-plaintext highlighter-rouge">read.csv()</code> usage, <code class="language-plaintext highlighter-rouge">read.csv()</code> is
a “version” of the function <code class="language-plaintext highlighter-rouge">read.table()</code> and accepts all its arguments.
If you set <code class="language-plaintext highlighter-rouge">dec=","</code> you could change the decimal operator. We’d probably
assume the delimiter is some other character.</p>

    <p>D) You can set <code class="language-plaintext highlighter-rouge">nrow</code> to a numeric value (e.g. <code class="language-plaintext highlighter-rouge">nrow=10000</code>) to choose how
many rows of a file you read in. This may be useful for very large files
where not all the data is needed to test some data cleaning steps you are
applying.</p>

    <p>Hopefully, this exercise gets you thinking about using the provided help
documentation in R. There are many arguments that exist, but which we wont
have time to cover. Look here to get familiar with functions you use
frequently, you may be surprised at what you find they can do.</p>
  </blockquote>
</blockquote>

<p>Now, let’s read in the file <code class="language-plaintext highlighter-rouge">combined_tidy_vcf.csv</code> which will be located in
<code class="language-plaintext highlighter-rouge">/home/dcuser/.solutions/R_data/</code>. Call this data <code class="language-plaintext highlighter-rouge">variants</code>. The
first argument to pass to our <code class="language-plaintext highlighter-rouge">read.csv()</code> function is the file path for our
data. The file path must be in quotes and now is a good time to remember to
use tab autocompletion. <strong>If you use tab autocompletion you avoid typos and
errors in file paths.</strong> Use it!</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## read in a CSV file and save it as 'variants'</span><span class="w">

</span><span class="n">variants</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="s2">"../r_data/combined_tidy_vcf.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>One of the first things you should notice is that in the Environment window,
you have the <code class="language-plaintext highlighter-rouge">variants</code> object, listed as 801 obs. (observations/rows)
of 29 variables (columns). Double-clicking on the name of the object will open
a view of the data in a new tab.</p>

<p><img src="../fig/rstudio_dataframeview.png" alt="rstudio data frame view" style="width: 1000px;" /></p>

<h2 id="summarizing-and-determining-the-structure-of-a-data-frame">Summarizing and determining the structure of a data frame.</h2>

<p>A <strong>data frame is the standard way in R to store tabular data</strong>. A data fame
could also be thought of as a collection of vectors, all of which have the same
length. Using only two functions, we can learn a lot about out data frame
including some summary statistics as well as well as the “structure” of the data
frame. Let’s examine what each of these functions can tell us:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## get summary statistics on a data frame</span><span class="w">

</span><span class="n">summary</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  sample_id            CHROM                POS             ID         
 Length:801         Length:801         Min.   :   1521   Mode:logical  
 Class :character   Class :character   1st Qu.:1115970   NA's:801      
 Mode  :character   Mode  :character   Median :2290361                 
                                       Mean   :2243682                 
                                       3rd Qu.:3317082                 
                                       Max.   :4629225                 
                                                                       
     REF                ALT                 QUAL          FILTER       
 Length:801         Length:801         Min.   :  4.385   Mode:logical  
 Class :character   Class :character   1st Qu.:139.000   NA's:801      
 Mode  :character   Mode  :character   Median :195.000                 
                                       Mean   :172.276                 
                                       3rd Qu.:225.000                 
                                       Max.   :228.000                 
                                                                       
   INDEL              IDV              IMF               DP       
 Mode :logical   Min.   : 2.000   Min.   :0.5714   Min.   : 2.00  
 FALSE:700       1st Qu.: 7.000   1st Qu.:0.8824   1st Qu.: 7.00  
 TRUE :101       Median : 9.000   Median :1.0000   Median :10.00  
                 Mean   : 9.396   Mean   :0.9219   Mean   :10.57  
                 3rd Qu.:11.000   3rd Qu.:1.0000   3rd Qu.:13.00  
                 Max.   :20.000   Max.   :1.0000   Max.   :79.00  
                 NA's   :700      NA's   :700                     
      VDB                 RPB              MQB              BQB        
 Min.   :0.0005387   Min.   :0.0000   Min.   :0.0000   Min.   :0.1153  
 1st Qu.:0.2180410   1st Qu.:0.3776   1st Qu.:0.1070   1st Qu.:0.6963  
 Median :0.4827410   Median :0.8663   Median :0.2872   Median :0.8615  
 Mean   :0.4926291   Mean   :0.6970   Mean   :0.5330   Mean   :0.7784  
 3rd Qu.:0.7598940   3rd Qu.:1.0000   3rd Qu.:1.0000   3rd Qu.:1.0000  
 Max.   :0.9997130   Max.   :1.0000   Max.   :1.0000   Max.   :1.0000  
                     NA's   :773      NA's   :773      NA's   :773     
      MQSB              SGB               MQ0F           ICB         
 Min.   :0.01348   Min.   :-0.6931   Min.   :0.00000   Mode:logical  
 1st Qu.:0.95494   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:801      
 Median :1.00000   Median :-0.6620   Median :0.00000                 
 Mean   :0.96428   Mean   :-0.6444   Mean   :0.01127                 
 3rd Qu.:1.00000   3rd Qu.:-0.6364   3rd Qu.:0.00000                 
 Max.   :1.01283   Max.   :-0.4536   Max.   :0.66667                 
 NA's   :48                                                          
   HOB                AC          AN        DP4                  MQ       
 Mode:logical   Min.   :1   Min.   :1   Length:801         Min.   :10.00  
 NA's:801       1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00  
                Median :1   Median :1   Mode  :character   Median :60.00  
                Mean   :1   Mean   :1                      Mean   :58.19  
                3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00  
                Max.   :1   Max.   :1                      Max.   :60.00  
                                                                          
    Indiv              gt_PL               gt_GT   gt_GT_alleles     
 Length:801         Length:801         Min.   :1   Length:801        
 Class :character   Class :character   1st Qu.:1   Class :character  
 Mode  :character   Mode  :character   Median :1   Mode  :character  
                                       Mean   :1                     
                                       3rd Qu.:1                     
                                       Max.   :1                     
                                                                     
</code></pre></div></div>

<p>Our data frame had 29 variables, so we get 29 fields that summarize the data.
The <code class="language-plaintext highlighter-rouge">QUAL</code>, <code class="language-plaintext highlighter-rouge">IMF</code>, and <code class="language-plaintext highlighter-rouge">VDB</code> variables (and several others) are
numerical data and so you get summary statistics on the min and max values for
these columns, as well as mean, median, and interquartile ranges. Many of the other variables
(e.g. <code class="language-plaintext highlighter-rouge">sample_id</code>) are treated as categorical data (which have special
treatment in R - more on this in a bit). The most frequent 6 different categories and the
number of times they appear (e.g. the sample_id called ‘SRR2584863’ appeared 25 times)
are displayed. There was only one value for <code class="language-plaintext highlighter-rouge">CHROM</code>, “CP000819.1” which appeared
in all 801 observations.</p>

<p>Before we operate on the data, we also need to know a little more about the
data frame structure to do that we use the <code class="language-plaintext highlighter-rouge">str()</code> function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## get the structure of a data frame</span><span class="w">

</span><span class="n">str</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>'data.frame':	801 obs. of  29 variables:
 $ sample_id    : chr  "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" ...
 $ CHROM        : chr  "CP000819.1" "CP000819.1" "CP000819.1" "CP000819.1" ...
 $ POS          : int  9972 263235 281923 433359 473901 648692 1331794 1733343 2103887 2333538 ...
 $ ID           : logi  NA NA NA NA NA NA ...
 $ REF          : chr  "T" "G" "G" "CTTTTTTT" ...
 $ ALT          : chr  "G" "T" "T" "CTTTTTTTT" ...
 $ QUAL         : num  91 85 217 64 228 210 178 225 56 167 ...
 $ FILTER       : logi  NA NA NA NA NA NA ...
 $ INDEL        : logi  FALSE FALSE FALSE TRUE TRUE FALSE ...
 $ IDV          : int  NA NA NA 12 9 NA NA NA 2 7 ...
 $ IMF          : num  NA NA NA 1 0.9 ...
 $ DP           : int  4 6 10 12 10 10 8 11 3 7 ...
 $ VDB          : num  0.0257 0.0961 0.7741 0.4777 0.6595 ...
 $ RPB          : num  NA 1 NA NA NA NA NA NA NA NA ...
 $ MQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
 $ BQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
 $ MQSB         : num  NA NA 0.975 1 0.916 ...
 $ SGB          : num  -0.556 -0.591 -0.662 -0.676 -0.662 ...
 $ MQ0F         : num  0 0.167 0 0 0 ...
 $ ICB          : logi  NA NA NA NA NA NA ...
 $ HOB          : logi  NA NA NA NA NA NA ...
 $ AC           : int  1 1 1 1 1 1 1 1 1 1 ...
 $ AN           : int  1 1 1 1 1 1 1 1 1 1 ...
 $ DP4          : chr  "0,0,0,4" "0,1,0,5" "0,0,4,5" "0,1,3,8" ...
 $ MQ           : int  60 33 60 60 60 60 60 60 60 60 ...
 $ Indiv        : chr  "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" ...
 $ gt_PL        : chr  "121,0" "112,0" "247,0" "91,0" ...
 $ gt_GT        : int  1 1 1 1 1 1 1 1 1 1 ...
 $ gt_GT_alleles: chr  "G" "T" "T" "CTTTTTTTT" ...
</code></pre></div></div>

<p>Ok, thats a lot up unpack! Some things to notice.</p>

<ul>
  <li>the object type <code class="language-plaintext highlighter-rouge">data.frame</code> is displayed in the first row along with its
dimensions, in this case 801 observations (rows) and 29 variables (columns)</li>
  <li>Each variable (column) has a name (e.g. <code class="language-plaintext highlighter-rouge">sample_id</code>). This is followed
by the object mode (e.g. factor, int, num, etc.). Notice that before each
variable name there is a <code class="language-plaintext highlighter-rouge">$</code> - this will be important later.</li>
</ul>

<h2 id="introducing-factors">Introducing Factors</h2>

<p>Factors are the final major data structure we will introduce in our R genomics
lessons. Factors can be thought of as vectors which are specialized for
categorical data. Given R’s specialization for statistics, this make sense since
categorial and continuous variables usually have different treatments. Sometimes
you may want to have data treated as a factor, but in other cases, this may be
undesirable.</p>

<p>Since some of the data in our data frame are factors, lets see how factors work. First, we’ll
extract one of the columns of our data frame to a new object, so that we don’t end up
modifying the <code class="language-plaintext highlighter-rouge">variants</code> object by mistake.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">## extract the "REF" column to a new object</span><span class="w">

</span><span class="n">REF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">variants</span><span class="o">$</span><span class="n">REF</span><span class="w">
</span></code></pre></div></div>

<p>Let’s look at the first few items in our factor using <code class="language-plaintext highlighter-rouge">head()</code>:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">REF</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "T"        "G"        "G"        "CTTTTTTT" "CCGC"     "C"       
</code></pre></div></div>

<p>What we get back are the items in our factor, and also something called “Levels”.
<strong>Levels are the different categories contained in a factor</strong>. By default, R
will organize the levels in a factor in alphabetical order. So the first level in this factor is
“A”.</p>

<p>Lets look at the contents of a factor in a slightly different way using <code class="language-plaintext highlighter-rouge">str()</code>:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">str</span><span class="p">(</span><span class="n">REF</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code> chr [1:801] "T" "G" "G" "CTTTTTTT" "CCGC" "C" "C" "G" ...
</code></pre></div></div>

<p>For the sake of efficiency, R stores the content of a factor as a vector of
integers, which an integer is assigned to each of the possible levels. Recall
levels are assigned in alphabetical order. In this case, the first item in our “REF” object is
“T”, which happens to be the 49th level of our factor, ordered alphabeticaly. The next two
items are both “G”s, which is the 33rd level of our factor.</p>

<h2 id="plotting-and-ordering-factors">Plotting and ordering factors</h2>

<p>One of the most common uses for factors will be when you plot categorical
values. For example, suppose we want to know how many of our variants had each possible 
nucleotide (or nucleotide combination) in the reference genome? We could generate a plot:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">plot</span><span class="p">(</span><span class="n">REF</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext warning highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Warning in xy.coords(x, y, xlabel, ylabel, log): NAs introduced by coercion
</code></pre></div></div>

<div class="language-plaintext warning highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Warning in min(x): no non-missing arguments to min; returning Inf
</code></pre></div></div>

<div class="language-plaintext warning highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Warning in max(x): no non-missing arguments to max; returning -Inf
</code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in plot.window(...): need finite 'ylim' values
</code></pre></div></div>

<p><img src="../fig/rmd-03-unnamed-chunk-9-1.png" title="plot of chunk unnamed-chunk-9" alt="plot of chunk unnamed-chunk-9" width="612" style="display: block; margin: auto;" /></p>

<p>This isn’t a particularly pretty example of a plot. We’ll be learning much more about creating nice, publication-quality graphics later in this lesson.</p>

<!-- For now, let's explore how we can order -->
<!-- the factors in our plot so that the first four values are "A", "C", "G", "T", with multi-nucleotide -->
<!-- combinations listed alphabetically after these four. -->

<!-- We can take our existing `REF` factor, and use the `factor()` -->
<!-- function again. This time we will pass it two new arguments: `levels` will be -->
<!-- assigned to a vector that has the REF values in the order we want them, -->
<!-- and we will set the `ordered` argument to TRUE. -->

<!-- ```{r, purl = FALSE} -->
<!--  # order the 'REF' factor to match our desired set of levels -->
<!-- REF <-  -->
<!-- ```         -->

<!-- We can now see the new ordering: -->

<!-- ```{r, purl = FALSE} -->

<!-- ``` -->

<!-- Although not all levels are shown, notice there are `<` signs indicating an -->
<!-- order. -->

<h2 id="subsetting-data-frames">Subsetting data frames</h2>

<p>Next, we are going to talk about how you can get specific values from data frames, and where necessary, change the mode of a column of values.</p>

<p>The first thing to remember is that a data frame is two-dimensional (rows and
columns). Therefore, to select a specific value we will will once again use
<code class="language-plaintext highlighter-rouge">[]</code> (bracket) notation, but we will specify more than one value (except in some cases
where we are taking a range).</p>

<blockquote class="challenge">
  <h2 id="exercise-subsetting-a-data-frame">Exercise: Subsetting a data frame</h2>

  <p><strong>Try the following indices and functions and try to figure out what they return</strong></p>

  <p>a. <code class="language-plaintext highlighter-rouge">variants[1,1]</code></p>

  <p>b. <code class="language-plaintext highlighter-rouge">variants[2,4]</code></p>

  <p>c. <code class="language-plaintext highlighter-rouge">variants[801,29]</code></p>

  <p>d. <code class="language-plaintext highlighter-rouge">variants[2, ]</code></p>

  <p>e. <code class="language-plaintext highlighter-rouge">variants[-1, ]</code></p>

  <p>f. <code class="language-plaintext highlighter-rouge">variants[1:4,1]</code></p>

  <p>g. <code class="language-plaintext highlighter-rouge">variants[1:10,c("REF","ALT")]</code></p>

  <p>h. <code class="language-plaintext highlighter-rouge">variants[,c("sample_id")]</code></p>

  <p>i. <code class="language-plaintext highlighter-rouge">head(variants)</code></p>

  <p>j. <code class="language-plaintext highlighter-rouge">tail(variants)</code></p>

  <p>k. <code class="language-plaintext highlighter-rouge">variants$sample_id</code></p>

  <p>l. <code class="language-plaintext highlighter-rouge">variants[variants$REF == "A",]</code></p>

  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>
    <p>a.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "SRR2584863"
</code></pre></div>    </div>

    <p>b.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="m">4</span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] NA
</code></pre></div>    </div>

    <p>c.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">801</span><span class="p">,</span><span class="m">29</span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "T"
</code></pre></div>    </div>

    <p>d.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   sample_id      CHROM    POS ID REF ALT QUAL FILTER INDEL IDV IMF DP      VDB
2 SRR2584863 CP000819.1 263235 NA   G   T   85     NA FALSE  NA  NA  6 0.096133
  RPB MQB BQB MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
2   1   1   1   NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
                                                               Indiv gt_PL
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
  gt_GT gt_GT_alleles
2     1             T
</code></pre></div>    </div>

    <p>e.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   sample_id      CHROM     POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
2 SRR2584863 CP000819.1  263235 NA        G         T   85     NA FALSE  NA  NA
3 SRR2584863 CP000819.1  281923 NA        G         T  217     NA FALSE  NA  NA
4 SRR2584863 CP000819.1  433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
5 SRR2584863 CP000819.1  473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
6 SRR2584863 CP000819.1  648692 NA        C         T  210     NA FALSE  NA  NA
7 SRR2584863 CP000819.1 1331794 NA        C         A  178     NA FALSE  NA  NA
  DP      VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
2  6 0.096133   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 10 0.774083  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 12 0.477704  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 10 0.659505  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 10 0.268014  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
7  8 0.624078  NA  NA  NA 0.900802 -0.651104 0.000000  NA  NA  1  1 0,0,3,5 60
                                                               Indiv gt_PL
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
7 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 208,0
  gt_GT gt_GT_alleles
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T
7     1             A
</code></pre></div>    </div>

    <p>f.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
</code></pre></div>    </div>

    <p>g.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s2">"REF"</span><span class="p">,</span><span class="s2">"ALT"</span><span class="p">)]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>                                REF
1                                 T
2                                 G
3                                 G
4                          CTTTTTTT
5                              CCGC
6                                 C
7                                 C
8                                 G
9  ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG
10                               AT
                                                        ALT
1                                                         G
2                                                         T
3                                                         T
4                                                 CTTTTTTTT
5                                                    CCGCGC
6                                                         T
7                                                         A
8                                                         A
9  ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG
10                                                      ATT
</code></pre></div>    </div>

    <p>h.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s2">"sample_id"</span><span class="p">)]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
[6] "SRR2584863"
</code></pre></div>    </div>

    <p>i.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>   sample_id      CHROM    POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
1 SRR2584863 CP000819.1   9972 NA        T         G   91     NA FALSE  NA  NA
2 SRR2584863 CP000819.1 263235 NA        G         T   85     NA FALSE  NA  NA
3 SRR2584863 CP000819.1 281923 NA        G         T  217     NA FALSE  NA  NA
4 SRR2584863 CP000819.1 433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
5 SRR2584863 CP000819.1 473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
6 SRR2584863 CP000819.1 648692 NA        C         T  210     NA FALSE  NA  NA
  DP       VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
1  4 0.0257451  NA  NA  NA       NA -0.556411 0.000000  NA  NA  1  1 0,0,0,4 60
2  6 0.0961330   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 10 0.7740830  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 12 0.4777040  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 10 0.6595050  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 10 0.2680140  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
                                                               Indiv gt_PL
1 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 121,0
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
  gt_GT gt_GT_alleles
1     1             G
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T
</code></pre></div>    </div>

    <p>j.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tail</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>     sample_id      CHROM     POS ID REF ALT QUAL FILTER INDEL IDV IMF DP
796 SRR2589044 CP000819.1 3444175 NA   G   T  184     NA FALSE  NA  NA  9
797 SRR2589044 CP000819.1 3481820 NA   A   G  225     NA FALSE  NA  NA 12
798 SRR2589044 CP000819.1 3893550 NA  AG AGG  101     NA  TRUE   4   1  4
799 SRR2589044 CP000819.1 3901455 NA   A  AC   70     NA  TRUE   3   1  3
800 SRR2589044 CP000819.1 4100183 NA   A   G  177     NA FALSE  NA  NA  8
801 SRR2589044 CP000819.1 4431393 NA TGG   T  225     NA  TRUE  10   1 10
          VDB RPB MQB BQB     MQSB       SGB MQ0F ICB HOB AC AN     DP4 MQ
796 0.4714620  NA  NA  NA 0.992367 -0.651104    0  NA  NA  1  1 0,0,4,4 60
797 0.8707240  NA  NA  NA 1.000000 -0.680642    0  NA  NA  1  1 0,0,4,8 60
798 0.9182970  NA  NA  NA 1.000000 -0.556411    0  NA  NA  1  1 0,0,3,1 52
799 0.0221621  NA  NA  NA       NA -0.511536    0  NA  NA  1  1 0,0,3,0 60
800 0.9272700  NA  NA  NA 0.900802 -0.651104    0  NA  NA  1  1 0,0,3,5 60
801 0.7488140  NA  NA  NA 1.007750 -0.670168    0  NA  NA  1  1 0,0,4,6 60
                                                                 Indiv gt_PL
796 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 214,0
797 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 255,0
798 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 131,0
799 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 100,0
800 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 207,0
801 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 255,0
    gt_GT gt_GT_alleles
796     1             T
797     1             G
798     1           AGG
799     1            AC
800     1             G
801     1             T
</code></pre></div>    </div>

    <p>k.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="o">$</span><span class="n">sample_id</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
[6] "SRR2584863"
</code></pre></div>    </div>

    <p>l.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">variants</span><span class="p">[</span><span class="n">variants</span><span class="o">$</span><span class="n">REF</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"A"</span><span class="p">,]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    sample_id      CHROM     POS ID REF ALT QUAL FILTER INDEL IDV IMF DP
11 SRR2584863 CP000819.1 2407766 NA   A   C  104     NA FALSE  NA  NA  9
12 SRR2584863 CP000819.1 2446984 NA   A   C  225     NA FALSE  NA  NA 20
14 SRR2584863 CP000819.1 2665639 NA   A   T  225     NA FALSE  NA  NA 19
16 SRR2584863 CP000819.1 3339313 NA   A   C  211     NA FALSE  NA  NA 10
18 SRR2584863 CP000819.1 3481820 NA   A   G  200     NA FALSE  NA  NA  9
19 SRR2584863 CP000819.1 3488669 NA   A   C  225     NA FALSE  NA  NA 13
         VDB      RPB      MQB      BQB     MQSB       SGB     MQ0F ICB HOB AC
11 0.0230738 0.900802 0.150134 0.750668 0.500000 -0.590765 0.333333  NA  NA  1
12 0.0714027       NA       NA       NA 1.000000 -0.689466 0.000000  NA  NA  1
14 0.9960390       NA       NA       NA 1.000000 -0.690438 0.000000  NA  NA  1
16 0.4059360       NA       NA       NA 1.007750 -0.670168 0.000000  NA  NA  1
18 0.1070810       NA       NA       NA 0.974597 -0.662043 0.000000  NA  NA  1
19 0.0162706       NA       NA       NA 1.000000 -0.680642 0.000000  NA  NA  1
   AN      DP4 MQ
11  1  3,0,3,2 25
12  1 0,0,10,6 60
14  1 0,0,12,5 60
16  1  0,0,4,6 60
18  1  0,0,4,5 60
19  1  0,0,8,4 60
                                                                Indiv gt_PL
11 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 131,0
12 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
14 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
16 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 241,0
18 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 230,0
19 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
   gt_GT gt_GT_alleles
11     1             C
12     1             C
14     1             T
16     1             C
18     1             G
19     1             C
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>The subsetting notation is very similar to what we learned for
vectors. The key differences include:</p>

<ul>
  <li>Typically provide two values separated by commas: data.frame[row, column]</li>
  <li>In cases where you are taking a continuous range of numbers use a colon
between the numbers (start:stop, inclusive)</li>
  <li>For a non continuous set of numbers, pass a vector using <code class="language-plaintext highlighter-rouge">c()</code></li>
  <li>Index using the name of a column(s) by passing them as vectors using <code class="language-plaintext highlighter-rouge">c()</code></li>
</ul>

<p>Finally, in all of the subsetting exercises above, we printed values to
the screen. You can create a new data frame object by assigning
them to a new object name:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># create a new data frame containing only observations from SRR2584863 </span><span class="w">

</span><span class="n">SRR2584863_variants</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">variants</span><span class="p">[</span><span class="n">variants</span><span class="o">$</span><span class="n">sample_id</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"SRR2584863"</span><span class="p">,]</span><span class="w">

</span><span class="c1"># check the dimension of the data frame</span><span class="w">

</span><span class="nf">dim</span><span class="p">(</span><span class="n">SRR2584863_variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 25 29
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># get a summary of the data frame</span><span class="w">

</span><span class="n">summary</span><span class="p">(</span><span class="n">SRR2584863_variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  sample_id            CHROM                POS             ID         
 Length:25          Length:25          Min.   :   9972   Mode:logical  
 Class :character   Class :character   1st Qu.:1331794   NA's:25       
 Mode  :character   Mode  :character   Median :2618472                 
                                       Mean   :2464989                 
                                       3rd Qu.:3488669                 
                                       Max.   :4616538                 
                                                                       
     REF                ALT                 QUAL         FILTER       
 Length:25          Length:25          Min.   : 31.89   Mode:logical  
 Class :character   Class :character   1st Qu.:104.00   NA's:25       
 Mode  :character   Mode  :character   Median :211.00                 
                                       Mean   :172.97                 
                                       3rd Qu.:225.00                 
                                       Max.   :228.00                 
                                                                      
   INDEL              IDV             IMF               DP      
 Mode :logical   Min.   : 2.00   Min.   :0.6667   Min.   : 2.0  
 FALSE:19        1st Qu.: 3.25   1st Qu.:0.9250   1st Qu.: 9.0  
 TRUE :6         Median : 8.00   Median :1.0000   Median :10.0  
                 Mean   : 7.00   Mean   :0.9278   Mean   :10.4  
                 3rd Qu.: 9.75   3rd Qu.:1.0000   3rd Qu.:12.0  
                 Max.   :12.00   Max.   :1.0000   Max.   :20.0  
                 NA's   :19      NA's   :19                     
      VDB               RPB              MQB               BQB        
 Min.   :0.01627   Min.   :0.9008   Min.   :0.04979   Min.   :0.7507  
 1st Qu.:0.07140   1st Qu.:0.9275   1st Qu.:0.09996   1st Qu.:0.7627  
 Median :0.37674   Median :0.9542   Median :0.15013   Median :0.7748  
 Mean   :0.40429   Mean   :0.9517   Mean   :0.39997   Mean   :0.8418  
 3rd Qu.:0.65951   3rd Qu.:0.9771   3rd Qu.:0.57507   3rd Qu.:0.8874  
 Max.   :0.99604   Max.   :1.0000   Max.   :1.00000   Max.   :1.0000  
                   NA's   :22       NA's   :22        NA's   :22      
      MQSB             SGB               MQ0F           ICB         
 Min.   :0.5000   Min.   :-0.6904   Min.   :0.00000   Mode:logical  
 1st Qu.:0.9599   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:25       
 Median :0.9962   Median :-0.6620   Median :0.00000                 
 Mean   :0.9442   Mean   :-0.6341   Mean   :0.04667                 
 3rd Qu.:1.0000   3rd Qu.:-0.6168   3rd Qu.:0.00000                 
 Max.   :1.0128   Max.   :-0.4536   Max.   :0.66667                 
 NA's   :3                                                          
   HOB                AC          AN        DP4                  MQ       
 Mode:logical   Min.   :1   Min.   :1   Length:25          Min.   :10.00  
 NA's:25        1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00  
                Median :1   Median :1   Mode  :character   Median :60.00  
                Mean   :1   Mean   :1                      Mean   :55.52  
                3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00  
                Max.   :1   Max.   :1                      Max.   :60.00  
                                                                          
    Indiv              gt_PL               gt_GT   gt_GT_alleles     
 Length:25          Length:25          Min.   :1   Length:25         
 Class :character   Class :character   1st Qu.:1   Class :character  
 Mode  :character   Mode  :character   Median :1   Mode  :character  
                                       Mean   :1                     
                                       3rd Qu.:1                     
                                       Max.   :1                     
                                                                     
</code></pre></div></div>

<h2 id="coercing-values-in-data-frames">Coercing values in data frames</h2>

<blockquote class="callout">
  <h2 id="tip-coercion-isnt-limited-to-data-frames">Tip: coercion isn’t limited to data frames</h2>

  <p>While we are going to address coercion in the context of data frames
most of these methods apply to other data structures, such as vectors</p>
</blockquote>

<p>Sometimes, it is possible that R will misinterpret the type of data represented
in a data frame, or store that data in a mode which prevents you from
operating on the data the way you wish. For example, a long list of gene names
isn’t usually thought of as a categorical variable, the way that your
experimental condition (e.g. control, treatment) might be. More importantly,
some R packages you use to analyze your data may expect characters as input,
not factors. At other times (such as plotting or some statistical analyses) a
factor may be more appropriate. Ultimately, you should know how to change the
mode of an object.</p>

<p>First, its very important to recognize that coercion happens in R all the time.
This can be a good thing when R gets it right, or a bad thing when the result
is not what you expect. Consider:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_chromosomes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s1">'3'</span><span class="p">,</span><span class="w"> </span><span class="s1">'11'</span><span class="p">,</span><span class="w"> </span><span class="s1">'X'</span><span class="p">,</span><span class="w"> </span><span class="s1">'6'</span><span class="p">)</span><span class="w">
</span><span class="n">typeof</span><span class="p">(</span><span class="n">snp_chromosomes</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "character"
</code></pre></div></div>

<p>Although there are several numbers in our vector, they are all in quotes, so
we have explicitly told R to consider them as characters. However, even if we removed
the quotes from the numbers, R would coerce everything into a character:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_chromosomes_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">11</span><span class="p">,</span><span class="w"> </span><span class="s1">'X'</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">)</span><span class="w">
</span><span class="n">typeof</span><span class="p">(</span><span class="n">snp_chromosomes_2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "character"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_chromosomes_2</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "3"
</code></pre></div></div>

<p>We can use the <code class="language-plaintext highlighter-rouge">as.</code> functions to explicitly coerce values from one form into
another. Consider the following vector of characters, which all happen to be
valid numbers:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_positions_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"8762685"</span><span class="p">,</span><span class="w"> </span><span class="s2">"66560624"</span><span class="p">,</span><span class="w"> </span><span class="s2">"67545785"</span><span class="p">,</span><span class="w"> </span><span class="s2">"154039662"</span><span class="p">)</span><span class="w">
</span><span class="n">typeof</span><span class="p">(</span><span class="n">snp_positions_2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "character"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_positions_2</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "8762685"
</code></pre></div></div>

<p>Now we can coerce <code class="language-plaintext highlighter-rouge">snp_positions_2</code> into a numeric type using <code class="language-plaintext highlighter-rouge">as.numeric()</code>:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_positions_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">snp_positions_2</span><span class="p">)</span><span class="w">
</span><span class="n">typeof</span><span class="p">(</span><span class="n">snp_positions_2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "double"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_positions_2</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 8762685
</code></pre></div></div>

<p>Sometimes coercion is straight forward, but what would happen if we tried
using <code class="language-plaintext highlighter-rouge">as.numeric()</code> on <code class="language-plaintext highlighter-rouge">snp_chromosomes_2</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_chromosomes_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">snp_chromosomes_2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext warning highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Warning: NAs introduced by coercion
</code></pre></div></div>

<p>If we check, we will see that an <code class="language-plaintext highlighter-rouge">NA</code> value (R’s default value for missing
data) has been introduced.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">snp_chromosomes_2</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1]  3 11 NA  6
</code></pre></div></div>

<p>Trouble can really start when we try to coerce a factor. For example, when we
try to coerce the <code class="language-plaintext highlighter-rouge">sample_id</code> column in our data frame into a numeric mode
look at the result:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">as.numeric</span><span class="p">(</span><span class="n">variants</span><span class="o">$</span><span class="n">sample_id</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext warning highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Warning: NAs introduced by coercion
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  [1] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
 [26] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
 [51] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
 [76] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[101] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[126] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[151] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[176] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[201] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[226] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[251] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[276] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[301] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[326] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[351] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[376] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[401] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[426] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[451] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[476] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[501] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[526] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[551] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[576] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[601] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[626] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[651] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[676] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[701] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[726] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[751] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[776] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[801] NA
</code></pre></div></div>

<p>Strangely, it works! Almost. Instead of giving an error message, R returns
numeric values, which in this case are the integers assigned to the levels in
this factor. This kind of behavior can lead to hard-to-find bugs, for example
when we do have numbers in a factor, and we get numbers from a coercion. If
we don’t look carefully, we may not notice a problem.</p>

<p>If you need to coerce an entire column you can overwrite it using an expression
like this one:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># make the 'REF' column a character type column</span><span class="w">
 
</span><span class="n">variants</span><span class="o">$</span><span class="n">REF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">variants</span><span class="o">$</span><span class="n">REF</span><span class="p">)</span><span class="w">

</span><span class="c1"># check the type of the column</span><span class="w">
</span><span class="n">typeof</span><span class="p">(</span><span class="n">variants</span><span class="o">$</span><span class="n">REF</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "character"
</code></pre></div></div>

<h2 id="stringsasfactors--false">StringsAsFactors = FALSE</h2>

<p>Lets summarize this section on coercion with a few take home messages.</p>

<ul>
  <li>When you explicitly coerce one data type into another (this is known as
<strong>explicit coercion</strong>), be careful to check the result. Ideally, you should try to see if its possible to avoid steps in your analysis that force you to
coerce.</li>
  <li>R will sometimes coerce without you asking for it. This is called
(appropriately) <strong>implicit coercion</strong>. For example when we tried to create
a vector with multiple data types, R chose one type through implicit
coercion.</li>
  <li>Check the structure (<code class="language-plaintext highlighter-rouge">str()</code>) of your data frames before working with them!</li>
</ul>

<p>Regarding the first bullet point, one way to avoid needless coercion when
importing a data frame using any one of the <code class="language-plaintext highlighter-rouge">read.table()</code> functions such as
<code class="language-plaintext highlighter-rouge">read.csv()</code> is to set the argument <code class="language-plaintext highlighter-rouge">StringsAsFactors</code> to FALSE. By default,
this argument is TRUE. Setting it to FALSE will treat any non-numeric column to
a character type. <code class="language-plaintext highlighter-rouge">read.csv()</code> documentation, you will also see you can
explicitly type your columns using the <code class="language-plaintext highlighter-rouge">colClasses</code> argument. Other R packages
(such as the Tidyverse “readr”) don’t have this particular conversion issue,
but many packages will still try to guess a  data type.</p>

<h2 id="data-frame-bonus-material-math-sorting-renaming">Data frame bonus material: math, sorting, renaming</h2>

<p>Here are a few operations that don’t need much explanation, but which are good
to know.</p>

<p>There are lots of arithmetic functions you may want to apply to your data
frame, covering those would be a course in itself (there is some starting
material <a href="https://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth/">here</a>). Our lessons will cover some additional summary statistical functions in
a subsequent lesson, but overall we will focus on data cleaning and
visualization.</p>

<p>You can use functions like <code class="language-plaintext highlighter-rouge">mean()</code>, <code class="language-plaintext highlighter-rouge">min()</code>, <code class="language-plaintext highlighter-rouge">max()</code> on an
individual column. Let’s look at the “DP” or filtered depth. This value shows the number of filtered
reads that support each of the reported variants.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">max</span><span class="p">(</span><span class="n">variants</span><span class="o">$</span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 79
</code></pre></div></div>

<p>You can sort a data frame using the <code class="language-plaintext highlighter-rouge">order()</code> function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sorted_by_DP</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">variants</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">variants</span><span class="o">$</span><span class="n">DP</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">sorted_by_DP</span><span class="o">$</span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 2 2 2 2 2 2
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>
  <p>The <code class="language-plaintext highlighter-rouge">order()</code> function lists values in increasing order by default. Look at the documentation
for this function and change <code class="language-plaintext highlighter-rouge">sorted_by_DP</code> to start with variants with the greatest filtered
depth (“DP”).</p>

  <blockquote class="solution">
    <h2 id="solution-2">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sorted_by_DP</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">variants</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">variants</span><span class="o">$</span><span class="n">DP</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">sorted_by_DP</span><span class="o">$</span><span class="n">DP</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 79 46 41 29 29 27
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<!-- You can selectively replace values in a data frame based on their value: -->

<!-- ```{r} -->
<!-- ``` -->

<p>You can rename columns:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">colnames</span><span class="p">(</span><span class="n">variants</span><span class="p">)[</span><span class="n">colnames</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"sample_id"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"strain"</span><span class="w">

</span><span class="c1"># check the column name (hint names are returned as a vector)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code> [1] "strain"        "CHROM"         "POS"           "ID"           
 [5] "REF"           "ALT"           "QUAL"          "FILTER"       
 [9] "INDEL"         "IDV"           "IMF"           "DP"           
[13] "VDB"           "RPB"           "MQB"           "BQB"          
[17] "MQSB"          "SGB"           "MQ0F"          "ICB"          
[21] "HOB"           "AC"            "AN"            "DP4"          
[25] "MQ"            "Indiv"         "gt_PL"         "gt_GT"        
[29] "gt_GT_alleles"
</code></pre></div></div>

<h2 id="saving-your-data-frame-to-a-file">Saving your data frame to a file</h2>

<p>We can save data to a file. We will save our <code class="language-plaintext highlighter-rouge">SRR2584863_variants</code> object
to a .csv file using the <code class="language-plaintext highlighter-rouge">write.csv()</code> function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">write.csv</span><span class="p">(</span><span class="n">SRR2584863_variants</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"../data/SRR2584863_variants.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">write.csv()</code> function has some additional arguments listed in the help, but
at a minimum you need to tell it what data frame to write to file, and give a
path to a file name in quotes (if you only provide a file name, the file will
be written in the current working directory).</p>

<h2 id="importing-data-from-excel">Importing data from Excel</h2>

<p>Excel is one of the most common formats, so we need to discuss how to make
these files play nicely with R. The simplest way to import data from Excel is
to <strong>save your Excel file in .csv format</strong>*. You can then import into R right
away. Sometimes you may not be able to do this (imagine you have data in 300
Excel files, are you going to open and export all of them?).</p>

<p>One common R package (a set of code with features you can download and add to
your R installation) is the <a href="https://CRAN.R-project.org/package=readxl">readxl package</a> which can open and import Excel
files. Rather than addressing package installation this second (we’ll discuss this soon!), we can take
advantage of RStudio’s import feature which integrates this package. (Note:
this feature is available only in the latest versions of RStudio such as is
installed on our cloud instance).</p>

<p>First, in the RStudio menu go to <strong>File</strong>, select <strong>Import Dataset</strong>, and
choose <strong>From Excel…</strong> (notice there are several other options you can
explore).</p>

<p><img src="../fig/rstudio_import_menu.png " alt="rstudio import menu" style="width: 600px;" /></p>

<p>Next, under <strong>File/Url:</strong> click the <kbd>Browse</kbd> button and navigate to the <strong>Ecoli_metadata.xlsx</strong> file located at <code class="language-plaintext highlighter-rouge">/home/dcuser/dc_sample_data/R</code>.
You should now see a preview of the data to be imported:</p>

<p><img src="../fig/rstudio_import_screen.png " alt="rstudio import screen" style="width: 1200px;" /></p>

<p>Notice that you have the option to change the data type of each variable by
clicking arrow (drop-down menu) next to each column title. Under <strong>Import
Options</strong> you may also rename the data, choose a different sheet to import, and
choose how you will handle headers and skipped rows. Under <strong>Code Preview</strong> you
can see the code that will be used to import this file. We could have written
this code and imported the Excel file without the RStudio import function, but
now you can choose your preference.</p>

<p>In this exercise, we will leave the title of the data frame as
<strong>Ecoli_metadata</strong>, and there are no other options we need to adjust. Click the
<kbd>Import</kbd> button to import the data.</p>

<p>Finally, let’s check the first few lines of the <code class="language-plaintext highlighter-rouge">Ecoli_metadata</code> data
frame:</p>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error: `path` does not exist: '../data/Ecoli_metadata.xlsx'
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in head(Ecoli_metadata): object 'Ecoli_metadata' not found
</code></pre></div></div>

<p>The type of this object is ‘tibble’, a type of data
frame we will talk more about in the ‘dplyr’ section. If you needed
a true R data frame you could coerce with <code class="language-plaintext highlighter-rouge">as.data.frame()</code>.</p>

<blockquote class="challenge">
  <h2 id="exercise-putting-it-all-together---data-frames">Exercise: Putting it all together - data frames</h2>

  <p><strong>Using the <code class="language-plaintext highlighter-rouge">Ecoli_metadata</code> data frame created above, answer the following questions</strong></p>

  <p>A) What are the dimensions (# rows, # columns) of the data frame?</p>

  <p>B) What are categories are there in the <code class="language-plaintext highlighter-rouge">cit</code> column? <em>hint</em>: treat column as factor</p>

  <p>C) How many of each of the <code class="language-plaintext highlighter-rouge">cit</code> categories are there?</p>

  <p>D) What is the genome size for the 7th observation in this data set?</p>

  <p>E) What is the median value of the variable <code class="language-plaintext highlighter-rouge">genome_size</code></p>

  <p>F) Rename the column <code class="language-plaintext highlighter-rouge">sample</code> to <code class="language-plaintext highlighter-rouge">sample_id</code></p>

  <p>G) Create a new column (name genome_size_bp) and set it equal to the genome_size multiplied by 1,000,000</p>

  <p>H) Save the edited Ecoli_metadata data frame as “exercise_solution.csv” in your current working directory.</p>

  <blockquote class="solution">
    <h2 id="solution-3">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">dim</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in eval(expr, envir, enclos): object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">levels</span><span class="p">(</span><span class="n">as.factor</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="o">$</span><span class="n">cit</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in is.factor(x): object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">table</span><span class="p">(</span><span class="n">as.factor</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="o">$</span><span class="n">cit</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in is.factor(x): object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">Ecoli_metadata</span><span class="p">[</span><span class="m">7</span><span class="p">,</span><span class="m">7</span><span class="p">]</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in eval(expr, envir, enclos): object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">median</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="o">$</span><span class="n">genome_size</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in median(Ecoli_metadata$genome_size): object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">colnames</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="p">)[</span><span class="n">colnames</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"sample"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"sample_id"</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in colnames(Ecoli_metadata)[colnames(Ecoli_metadata) == "sample"] &lt;- "sample_id": object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">Ecoli_metadata</span><span class="o">$</span><span class="n">genome_size_bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Ecoli_metadata</span><span class="o">$</span><span class="n">genome_size</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">1000000</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in eval(expr, envir, enclos): object 'Ecoli_metadata' not found
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">write.csv</span><span class="p">(</span><span class="n">Ecoli_metadata</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"exercise_solution.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in is.data.frame(x): object 'Ecoli_metadata' not found
</code></pre></div>    </div>
  </blockquote>
</blockquote>





<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>It is easy to import data into R from tabular formats including Excel. However, you still need to check that R has imported and interpreted your data correctly</p>
</li>
    
    <li><p>There are best practices for organizing your data (keeping it tidy) and R is great for this</p>
</li>
    
    <li><p>Base R has many useful functions for manipulating your data, but all of R’s capabilities are greatly enhanced by software packages developed by the community</p>
</li>
    
  </ul>
</blockquote>

</article>



























  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../02-r-basics/index.html"><span class="glyphicon glyphicon-menu-left" aria-hidden="true"></span><span class="sr-only">previous episode</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../04-dplyr/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>



      
      






<footer>
  <hr/>
  <div class="row">
    <div class="col-md-6 license" id="license-info" align="left">
	
	Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2018–2020
	by <a href="https://carpentries.org/">The Carpentries</a>
        <br>
        Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2016–2018
	by <a href="http://datacarpentry.org">Data Carpentry</a>
	
    </div>
    <div class="col-md-6 help-links" align="right">
	
	
	<a href="/edit//_episodes_rmd/03-basics-factors-dataframes.Rmd" data-checker-ignore>Edit on GitHub</a>
	
	
	/
	<a href="/blob//CONTRIBUTING.md" data-checker-ignore>Contributing</a>
	/
	<a href="/">Source</a>
	/
	<a href="/blob//CITATION" data-checker-ignore>Cite</a>
	/
	<a href="mailto:team@carpentries.org">Contact</a>
    </div>
  </div>
  <p class="text-muted text-right">
    <small><i>Using <a href="https://github.com/carpentries/carpentries-theme/">The Carpentries theme</a> &mdash; Site last built on: 2020-12-18 15:56:59 +0000.</i></small>
  </p>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-37305346-2', 'auto');
  ga('send', 'pageview');
</script>

  </body>
</html>
